app/src/main/java/com/firebase/uidemo/storage/ImageActivity.java

```diff
@@ -0,0 +1,163 @@
+package com.firebase.uidemo.storage;
+
+import android.Manifest;
+import android.content.Intent;
+import android.net.Uri;
+import android.os.Bundle;
+import android.provider.MediaStore;
+import android.support.annotation.NonNull;
+import android.support.v7.app.AppCompatActivity;
+import android.util.Log;
+import android.view.View;
+import android.widget.Button;
+import android.widget.ImageView;
+import android.widget.Toast;
+
+import com.bumptech.glide.Glide;
+import com.firebase.ui.storage.images.FirebaseImageLoader;
+import com.firebase.uidemo.R;
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.auth.AuthResult;
+import com.google.firebase.auth.FirebaseAuth;
+import com.google.firebase.storage.FirebaseStorage;
+import com.google.firebase.storage.StorageReference;
+import com.google.firebase.storage.UploadTask;
+
+import java.util.UUID;
+
+import butterknife.BindView;
+import butterknife.ButterKnife;
+import butterknife.OnClick;
+import pub.devrel.easypermissions.AfterPermissionGranted;
+import pub.devrel.easypermissions.EasyPermissions;
+
+public class ImageActivity extends AppCompatActivity {
+
+    private static final String TAG = "ImageDemo";
+    private static final int RC_CHOOSE_PHOTO = 101;
+    private static final int RC_IMAGE_PERMS = 102;
+
+    private StorageReference mImageRef;
+
+    @BindView(R.id.button_choose_photo)
+    Button mUploadButton;
+
+    @BindView(R.id.button_download_direct)
+    Button mDownloadDirectButton;
+
+    @BindView(R.id.first_image)
+    ImageView mImageView;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_image);
+        ButterKnife.bind(this);
+
+        // Authenticate so that uploading/downloading works
```

Comment:
Since authenticated downloads is a feature of the lib I'd like to keep auth, but I will add a comment here to explain and log some better resolution steps.